File: Algorithms/Implementation/3D Surface Area/solution.js
// https://www.hackerrank.com/challenges/3d-surface-area/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function surfaceArea(A) {
    var area = 0;

    for(var i = 0; i < A.length; i++){
        for(var j = 0; j < A[i].length; j++){
            area += 4 * A[i][j] + 2;

            if(i + 1 < A.length)
                area -= 2 * Math.min(A[i][j], A[i + 1][j]);

            if(j + 1 < A[i].length)
                    area -= 2 * Math.min(A[i][j], A[i][j + 1]);
        }   
    }

    return area;

}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const HW = readLine().split(' ');

    const H = parseInt(HW[0], 10);

    const W = parseInt(HW[1], 10);

    let A = Array(H);

    for (let i = 0; i < H; i++) {
        A[i] = readLine().split(' ').map(ATemp => parseInt(ATemp, 10));
    }

    let result = surfaceArea(A);

    ws.write(result + "\n");

    ws.end();
}

File: Algorithms/Implementation/3D Surface Area/solution.cs
// https://www.hackerrank.com/challenges/3d-surface-area/problem
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

class Solution {

    static int surfaceArea(int[][] A) {
        int area = 0;

        for(int i = 0; i < A.Length; i++){
            for(int j = 0; j < A[i].Length; j++){
                area += 4 * A[i][j] + 2;

                if(i + 1 < A.Length)
                    area -= 2 * Math.Min(A[i][j], A[i + 1][j]);

                if(j + 1 < A[i].Length)
                    area -= 2 * Math.Min(A[i][j], A[i][j + 1]);
            }   
        }

        return area;

    }

    static void Main(string[] args) {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        string[] HW = Console.ReadLine().Split(' ');

        int H = Convert.ToInt32(HW[0]);

        int W = Convert.ToInt32(HW[1]);

        int[][] A = new int[H][];

        for (int i = 0; i < H; i++) {
            A[i] = Array.ConvertAll(Console.ReadLine().Split(' '), ATemp => Convert.ToInt32(ATemp));
        }

        int result = surfaceArea(A);

        textWriter.WriteLine(result);

        textWriter.Flush();
        textWriter.Close();
    }
}
